0029-loopback-transport/0029-loopback-transport.md

```diff
@@ -0,0 +1,33 @@
+1. The sender's LT module is initialized with this open connection.
+1. The sender's LT module discovers the loopback address using [IL-DCP](https://github.com/interledgerjs/ilp-protocol-ildcp).
+1. The sender's application layer protocol tells the LT module to send a payment for a given source amount, and expiry time.
+1. The sender's LT module chooses a fulfillment at random.
+1. The sender's LT module creates the condition of their payment by computing the SHA-256 hash of the fulfillment.
+1. The sender's LT module sets the destination address to the loopback address for this lookback link.
+1. Using this condition, address, and amount, the sender's LT module prepares an ILPv4 payment to a connector of her choice.
+1. The payment will find its way from this connector, via any number of additional connectors, to the receiver.
+1. The receiver checks that the destination address matches the loopback address he announced to the sender over IL-DCP earlier, and forwards the ILPv4 `prepare` message to the sender, over the loopback link.
+1. The sender's application layer protocol gets to decide whether the destination amount that looped back is sufficient and whether this particular loopback payment should be accepted or rejected.
+1. The sender's LT module accepts or rejects the ILPv4 payment that looped back over the loopback link.
+1. If for whatever reason the payment does not arrive in time, the LT module responds to the application layer protocol with an error message instead
+
+## How to use
+
+When designing an application layer protocol, it's advisable to support both LT and PSK. Even though LT is far simpler than PSK, the extra loopback hop means that LT payments inherently have a higher latency than PSK payments, where the receiver can immediately fulfill incoming prepared payments. In situations where the sender can easily formulate instructions for the receiver about when to accept a payment and when not, and the sender trusts the receiver to execute these instructions in good faith, the speed up achieved with PSK means money needs to be tied up for less time, and this may translate to lower transaction fees, depending on connector pricing policies. So the application layer protocol needs to provide a choice between LT and PSK.
+
+If LT is chosen, the loopback link needs to be established; this will usually require some data exchange (e.g., one of the two parties sends connection details to the other party).
+
+The flow described above assumes one party only acts as the sender, and the other party only acts as the receiver, but once the loopback link is established, it can easily be used in both directions (in that case, each participant would send an IL-DCP request to the other participant, to ask for a loopback address).
```

Comment:
Only if the spec says that's what the implementations need to support. Can the receiver assume that the client has implemented ILDCP?

Reply:
i guess they can just try sending an IL-DCP request and see if they get a reply. any extra protocol for signalling whether that request will probably fail will not really add much information?

Reply:
I'm not suggesting another protocol for signaling whether it supports IL-DCP, my point is that this spec should specify what is expected of implementations and what others can rely on. If I implement it such that it awaits an IL-DCP call and throws an error if it doesn't hear anything back, have I implemented this spec (im)properly?

Reply:
You're right, I should probably leave the specific way in which the sender discovers the loopback address out of scope. For instance, it could be that the sender announces their own loopback address to the receiver using a route broadcast, or that their ILPv4 link uses hard-configured addresses.

Reply:
Hm, after trying that in https://github.com/interledger/rfcs/pull/416/commits/08e86ac2d68d46165dd7827a23e79b5530627aa8 I reverted it again because I realized it only makes the setup harder. So let's say the receiver always supports IL-DCP, and the sender always uses IL-DCP to discover the loopback address.

So then yes, the sender can throw an error if the IL-DCP call fails.